File: scripts/carga-historica.mjs
/**
 * Script de carga histórica — busca notícias dos últimos 30 dias em todos os feeds
 * Uso: node scripts/carga-historica.mjs [dias]
 */
import { XMLParser } from "fast-xml-parser";
import axios from "axios";
import mysql from "mysql2/promise";
import dotenv from "dotenv";
dotenv.config();

const DIAS = parseInt(process.argv[2] || "30");

const FEEDS = [
  { nome: "Poder360",       url: "https://www.poder360.com.br/feed/" },
  { nome: "Folha",          url: "https://feeds.folha.uol.com.br/poder/rss091.xml" },
  { nome: "G1",             url: "https://g1.globo.com/dynamo/politica/rss2.xml" },
  { nome: "G1 Brasil",      url: "https://g1.globo.com/dynamo/brasil/rss2.xml" },
  { nome: "CNN Brasil",     url: "https://www.cnnbrasil.com.br/feed/" },
  { nome: "Metrópoles",     url: "https://www.metropoles.com/feed" },
  { nome: "Veja",           url: "https://veja.abril.com.br/feed/" },
  { nome: "Exame",          url: "https://exame.com/feed/" },
  { nome: "Gazeta do Povo", url: "https://www.gazetadopovo.com.br/feed/politica/" },
  { nome: "Agência Brasil", url: "https://agenciabrasil.ebc.com.br/rss/politica/feed.xml" },
];

const TERMOS_PESQUISA = [
  "pesquisa", "sondagem", "instituto", "datafolha", "quaest", "ipec",
  "ibope", "atlas intel", "paraná pesquisas", "real time big data", "btg",
  "genial inteligência", "mda", "opinion box", "levantamento",
  "intenção de voto", "aprovação", "rejeição", "popularidade",
  "pesquisa eleitoral", "pesquisa de opinião", "pesquisa presidencial",
  "pesquisa aponta", "pesquisa mostra", "pesquisa revela", "pesquisa indica",
  "nova pesquisa", "última pesquisa", "pesquisa divulgada", "pesquisa registra",
  "cenário eleitoral", "corrida eleitoral", "disputa eleitoral",
  "primeiro turno", "segundo turno", "intenções de voto",
];

const TERMOS_ELEITORAIS = [
  "eleição", "eleitoral", "candidato", "candidatura", "presidente",
  "presidencial", "governador", "senador", "senado", "câmara", "deputado",
  "voto", "urna", "2026", "eleições 2026", "campanha eleitoral",
  "lula", "bolsonaro", "tarcísio", "tarcisio", "marçal", "pablo marçal",
  "ratinho junior", "ciro gomes", "alckmin", "eleitorado",
  "pré-candidato", "pré-candidatura", "disputa presidencial",
  "corrida presidencial", "pleito",
];

function contemTermo(texto, termos) {
  const t = texto.toLowerCase();
  return termos.some(termo => t.includes(termo));
}

function filtroValido(titulo, resumo) {
  const texto = `${titulo} ${resumo}`;
  return contemTermo(texto, TERMOS_PESQUISA) && contemTermo(texto, TERMOS_ELEITORAIS);
}

function detectarCategoria(texto) {
  const t = texto.toLowerCase();
  if (t.includes("presidente") || t.includes("presidencial") || t.includes("presidência") ||
      t.includes("lula") || t.includes("bolsonaro") || t.includes("tarcísio") ||
      t.includes("tarcisio") || t.includes("marçal") || t.includes("ciro gomes")) return "presidente";
  if (t.includes("governador") || t.includes("governo estadual") || t.includes("ratinho")) return "governador";
  if (t.includes("senador") || t.includes("senado")) return "senador";
  return "geral";
}

function limparTexto(texto) {
  if (!texto) return "";
  return texto
    .replace(/<!\[CDATA\[([\s\S]*?)\]\]>/g, "$1")
    .replace(/<[^>]+>/g, " ")
    .replace(/&amp;/g, "&").replace(/&lt;/g, "<").replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"').replace(/&#39;/g, "'").replace(/&nbsp;/g, " ")
    .replace(/\s+/g, " ").trim();
}

async function parsearFeed(feedUrl, nomeFonte) {
  try {
    const resp = await axios.get(feedUrl, {
      timeout: 20000,
      headers: {
        "User-Agent": "Mozilla/5.0 (compatible; PainelEleitoral/1.0)",
        "Accept": "application/rss+xml, application/xml, text/xml, */*",
      },
      responseType: "text",
    });

    const parser = new XMLParser({
      ignoreAttributes: false,
      attributeNamePrefix: "@_",
      cdataPropName: "__cdata",
      parseTagValue: false,
      trimValues: true,
      processEntities: true,
      htmlEntities: true,
    });

    const parsed = parser.parse(resp.data);
    const channel = parsed?.rss?.channel || parsed?.feed || parsed?.["rdf:RDF"]?.channel || null;
    if (!channel) return [];

    const items = Array.isArray(channel.item) ? channel.item
      : channel.item ? [channel.item]
      : Array.isArray(channel.entry) ? channel.entry
      : channel.entry ? [channel.entry] : [];

    return items.map(item => {
      const titulo = limparTexto(item.title?.__cdata || (typeof item.title === 'string' ? item.title : '') || "");
      // Extrai link de diferentes formatos (RSS 2.0, Atom, RDF)
      let link = "";
      if (typeof item.link === "string") link = item.link;
      else if (item.link?.__cdata) link = item.link.__cdata;
      else if (item.link?.['@_href']) link = item.link['@_href'];
      else if (Array.isArray(item.link)) {
        const alt = item.link.find(l => l['@_rel'] === 'alternate' || !l['@_rel']);
        link = alt?.['@_href'] || item.link[0]?.['@_href'] || "";
      }
      if (!link) link = item.guid?.__cdata || (typeof item.guid === "string" ? item.guid : "") || "";
      const descricao = limparTexto(
        item.description?.__cdata || item.description ||
        item.summary?.__cdata || item.summary ||
        item["content:encoded"]?.__cdata || ""
      );
      const pubDate = item.pubDate || item.published || item.updated || item["dc:date"] || null;
      let dataPublicacao = null;
      if (pubDate) { const d = new Date(pubDate); if (!isNaN(d.getTime())) dataPublicacao = d; }
      return { titulo, url: link.trim(), resumo: descricao.slice(0, 500), dataPublicacao };
    }).filter(i => i.titulo && i.url);
  } catch (err) {
    console.error(`  ✗ Erro ao buscar ${nomeFonte}: ${err.message}`);
    return [];
  }
}

// ─── Conexão com banco ────────────────────────────────────────────────────────
const dbUrl = process.env.DATABASE_URL;
if (!dbUrl) { console.error("DATABASE_URL não definida"); process.exit(1); }

const url = new URL(dbUrl);
const conn = await mysql.createPool({
  host: url.hostname,
  port: parseInt(url.port || "3306"),
  user: url.username,
  password: url.password,
  database: url.pathname.replace("/", ""),
  charset: "UTF8MB4_UNICODE_CI",
  ssl: { rejectUnauthorized: true },
  waitForConnections: true,
  connectionLimit: 5,
});

// ─── Execução principal ───────────────────────────────────────────────────────
const dataLimite = new Date();
dataLimite.setDate(dataLimite.getDate() - DIAS);
console.log(`\n🔍 Buscando notícias dos últimos ${DIAS} dias (desde ${dataLimite.toLocaleDateString("pt-BR")})\n`);

let totalEncontradas = 0;
let totalInseridas = 0;
const porFonte = {};

for (const feed of FEEDS) {
  process.stdout.write(`  Buscando ${feed.nome}... `);
  const items = await parsearFeed(feed.url, feed.nome);

  const relevantes = items.filter(item => {
    if (item.dataPublicacao && item.dataPublicacao < dataLimite) return false;
    return filtroValido(item.titulo, item.resumo);
  });

  console.log(`${items.length} itens → ${relevantes.length} relevantes`);
  totalEncontradas += relevantes.length;

  for (const item of relevantes) {
    const categoria = detectarCategoria(`${item.titulo} ${item.resumo}`);
    try {
      const [result] = await conn.execute(
        `INSERT IGNORE INTO noticias (titulo, url, fonte, data_publicacao, resumo, categoria)
         VALUES (?, ?, ?, ?, ?, ?)`,
        [item.titulo, item.url, feed.nome, item.dataPublicacao ?? new Date(), item.resumo, categoria]
      );
      if (result.affectedRows > 0) {
        totalInseridas++;
        porFonte[feed.nome] = (porFonte[feed.nome] || 0) + 1;
      }
    } catch (err) {
      // Silencia duplicatas e erros menores
    }
  }
}

// Registrar no histórico
const fontesList = Object.entries(porFonte).map(([k, v]) => `${k}: ${v}`).join(", ");
await conn.execute(
  `INSERT INTO atualizacoes (tipo, descricao, qtd_inseridas) VALUES (?, ?, ?)`,
  ["carga_historica", `Carga histórica ${DIAS}d — ${totalEncontradas} encontradas, ${totalInseridas} novas${fontesList ? ` (${fontesList})` : ""}`, totalInseridas]
);

await conn.end();

console.log(`\n✅ Carga histórica concluída!`);
console.log(`   Total encontradas: ${totalEncontradas}`);
console.log(`   Novas inseridas:   ${totalInseridas}`);
if (Object.keys(porFonte).length > 0) {
  console.log(`   Por fonte:`);
  for (const [fonte, qtd] of Object.entries(porFonte)) {
    console.log(`     ${fonte}: ${qtd}`);
  }
}

File: scripts/carga_historica.py
#!/usr/bin/env python3
"""
Carga histórica de notícias eleitorais via RSS
Uso: python3 scripts/carga_historica.py [dias]

Estratégia de filtro em duas camadas:
  1. Filtro PESQUISA: notícias com termos de pesquisa/sondagem + termos eleitorais (foco principal)
  2. Filtro ELEITORAL: notícias sobre eleições 2026, candidaturas, campanhas (contexto amplo)
"""
import sys
import re
import os
import feedparser
import mysql.connector
from datetime import datetime, timedelta, timezone
from html.parser import HTMLParser
from urllib.parse import urlparse

DIAS = int(sys.argv[1]) if len(sys.argv) > 1 else 30

FEEDS = [
    {"nome": "Poder360",       "url": "https://www.poder360.com.br/feed/"},
    {"nome": "Folha",          "url": "https://feeds.folha.uol.com.br/poder/rss091.xml"},
    {"nome": "G1",             "url": "https://g1.globo.com/dynamo/politica/rss2.xml"},
    {"nome": "G1 Brasil",      "url": "https://g1.globo.com/dynamo/brasil/rss2.xml"},
    {"nome": "CNN Brasil",     "url": "https://www.cnnbrasil.com.br/feed/"},
    {"nome": "Metrópoles",     "url": "https://www.metropoles.com/feed"},
    {"nome": "Veja",           "url": "https://veja.abril.com.br/feed/"},
    {"nome": "Exame",          "url": "https://exame.com/feed/"},
    {"nome": "Gazeta do Povo", "url": "https://www.gazetadopovo.com.br/feed/politica/"},
    {"nome": "Agência Brasil", "url": "https://agenciabrasil.ebc.com.br/rss/politica/feed.xml"},
]

# ─── Termos de pesquisa eleitoral ────────────────────────────────────────────
TERMOS_PESQUISA = [
    "pesquisa", "sondagem", "instituto", "datafolha", "quaest", "ipec",
    "ibope", "atlas intel", "paraná pesquisas", "real time big data", "btg",
    "genial inteligência", "mda", "opinion box", "levantamento",
    "intenção de voto", "intenções de voto", "aprovação", "rejeição",
    "popularidade", "pesquisa eleitoral", "pesquisa de opinião",
    "pesquisa presidencial", "pesquisa aponta", "pesquisa mostra",
    "pesquisa revela", "pesquisa indica", "nova pesquisa", "última pesquisa",
    "pesquisa divulgada", "pesquisa registra", "cenário eleitoral",
    "primeiro turno", "segundo turno", "votos", "% nas pesquisas",
    "lidera", "empate técnico", "margem de erro",
]

# ─── Termos eleitorais (para filtro duplo com pesquisa) ──────────────────────
TERMOS_ELEITORAIS_PESQUISA = [
    "eleição", "eleitoral", "candidato", "presidente", "presidencial",
    "governador", "senador", "senado", "câmara", "deputado",
    "voto", "urna", "2026", "eleições 2026", "lula", "bolsonaro",
    "tarcísio", "tarcisio", "marçal", "pablo marçal", "ratinho",
    "ciro gomes", "alckmin", "eleitorado", "corrida presidencial",
    "disputa presidencial", "pleito",
]

# ─── Termos eleitorais amplos (notícias de contexto eleitoral) ───────────────
TERMOS_ELEITORAIS_AMPLOS = [
    "eleições 2026", "eleição 2026", "campanha eleitoral", "campanha de 2026",
    "pré-candidato", "pré-candidatura", "candidatura à presidência",
    "candidatura ao governo", "candidatura ao senado",
    "corrida presidencial", "disputa presidencial", "disputa ao governo",
    "corrida ao governo", "chapa eleitoral", "coligação eleitoral",
    "filiação partidária", "convenção partidária", "janela partidária",
    "reforma eleitoral", "calendário eleitoral", "tse", "tribunal superior eleitoral",
    "urna eletrônica", "sistema eleitoral", "voto impresso",
    "debate presidencial", "debate eleitoral",
]

# ─── Termos de candidatos e partidos (contexto eleitoral direto) ─────────────
TERMOS_CANDIDATOS = [
    "lula 2026", "bolsonaro 2026", "tarcísio de freitas", "tarcisio de freitas",
    "pablo marçal", "ratinho junior 2026", "ciro gomes 2026",
    "simone tebet 2026", "romário 2026", "flávio bolsonaro 2026",
    "pt 2026", "pl 2026", "novo 2026", "psol 2026", "mdb 2026",
]


class HTMLStripper(HTMLParser):
    def __init__(self):
        super().__init__()
        self.reset()
        self.fed = []

    def handle_data(self, d):
        self.fed.append(d)

    def get_data(self):
        return " ".join(self.fed)


def limpar_html(texto):
    if not texto:
        return ""
    s = HTMLStripper()
    try:
        s.feed(str(texto))
        return re.sub(r"\s+", " ", s.get_data()).strip()
    except Exception:
        return re.sub(r"<[^>]+>", " ", str(texto)).strip()


def contem_termo(texto, termos):
    t = texto.lower()
    return any(termo in t for termo in termos)


def filtro_valido(titulo, resumo):
    """
    Aceita a notícia se:
    1. Contém termos de pesquisa E termos eleitorais (filtro principal), OU
    2. Contém termos eleitorais amplos (candidaturas, campanhas 2026), OU
    3. Contém termos de candidatos específicos 2026
    """
    texto = f"{titulo} {resumo}"
    # Filtro 1: pesquisa + eleitoral
    if contem_termo(texto, TERMOS_PESQUISA) and contem_termo(texto, TERMOS_ELEITORAIS_PESQUISA):
        return True
    # Filtro 2: contexto eleitoral amplo
    if contem_termo(texto, TERMOS_ELEITORAIS_AMPLOS):
        return True
    # Filtro 3: candidatos 2026
    if contem_termo(texto, TERMOS_CANDIDATOS):
        return True
    return False


def detectar_categoria(texto):
    t = texto.lower()
    if any(x in t for x in ["presidente", "presidencial", "presidência", "lula",
                              "bolsonaro", "tarcísio", "tarcisio", "marçal",
                              "ciro gomes", "disputa presidencial", "corrida presidencial",
                              "candidatura à presidência"]):
        return "presidente"
    if any(x in t for x in ["governador", "governo estadual", "governo do estado",
                              "ratinho", "candidatura ao governo", "disputa ao governo"]):
        return "governador"
    if any(x in t for x in ["senador", "senado", "vaga no senado",
                              "candidatura ao senado"]):
        return "senador"
    return "geral"


def parse_data(entry):
    for campo in ["published_parsed", "updated_parsed", "created_parsed"]:
        val = getattr(entry, campo, None)
        if val:
            try:
                return datetime(*val[:6], tzinfo=timezone.utc)
            except Exception:
                pass
    return None


def extrair_link(entry):
    if hasattr(entry, "link") and entry.link:
        return entry.link
    if hasattr(entry, "links"):
        for l in entry.links:
            if l.get("rel") == "alternate":
                return l.get("href", "")
    if hasattr(entry, "id") and entry.id and entry.id.startswith("http"):
        return entry.id
    return ""


# ─── Conexão MySQL ────────────────────────────────────────────────────────────
db_url = os.environ.get("DATABASE_URL", "")
if not db_url:
    print("❌ DATABASE_URL não definida")
    sys.exit(1)

parsed_url = urlparse(db_url)
conn = mysql.connector.connect(
    host=parsed_url.hostname,
    port=parsed_url.port or 3306,
    user=parsed_url.username,
    password=parsed_url.password,
    database=parsed_url.path.lstrip("/"),
    charset="utf8mb4",
    collation="utf8mb4_unicode_ci",
    ssl_verify_cert=False,
    ssl_verify_identity=False,
)
cursor = conn.cursor()

# ─── Execução principal ───────────────────────────────────────────────────────
data_limite = datetime.now(timezone.utc) - timedelta(days=DIAS)
print(f"\n🔍 Buscando notícias dos últimos {DIAS} dias (desde {data_limite.strftime('%d/%m/%Y')})\n")

total_encontradas = 0
total_inseridas = 0
por_fonte = {}

feedparser.USER_AGENT = "Mozilla/5.0 (compatible; PainelEleitoral/1.0)"

for feed in FEEDS:
    print(f"  Buscando {feed['nome']}... ", end="", flush=True)
    try:
        parsed = feedparser.parse(feed["url"])
        entries = parsed.entries
        print(f"{len(entries)} itens", end="")

        relevantes = []
        for entry in entries:
            data = parse_data(entry)
            if data and data < data_limite:
                continue

            titulo = limpar_html(getattr(entry, "title", ""))
            resumo_raw = ""
            if hasattr(entry, "content") and entry.content:
                resumo_raw = entry.content[0].get("value", "")
            if not resumo_raw:
                resumo_raw = getattr(entry, "summary", "") or getattr(entry, "description", "")
            resumo = limpar_html(resumo_raw)[:500]
            url = extrair_link(entry)

            if not titulo or not url:
                continue
            if filtro_valido(titulo, resumo):
                relevantes.append({
                    "titulo": titulo,
                    "url": url,
                    "resumo": resumo,
                    "data": data or datetime.now(timezone.utc),
                })

        print(f" → {len(relevantes)} relevantes")
        total_encontradas += len(relevantes)

        for item in relevantes:
            categoria = detectar_categoria(f"{item['titulo']} {item['resumo']}")
            try:
                cursor.execute(
                    """INSERT IGNORE INTO noticias (titulo, url, fonte, data_publicacao, resumo, categoria)
                       VALUES (%s, %s, %s, %s, %s, %s)""",
                    (item["titulo"], item["url"], feed["nome"],
                     item["data"].replace(tzinfo=None), item["resumo"], categoria)
                )
                if cursor.rowcount > 0:
                    total_inseridas += 1
                    por_fonte[feed["nome"]] = por_fonte.get(feed["nome"], 0) + 1
            except Exception:
                pass

        conn.commit()

    except Exception as e:
        print(f"\n  ✗ Erro: {e}")

# Registrar no histórico
fontes_list = ", ".join(f"{k}: {v}" for k, v in sorted(por_fonte.items(), key=lambda x: -x[1]))
descricao = f"Carga histórica {DIAS}d — {total_encontradas} encontradas, {total_inseridas} novas"
if fontes_list:
    descricao += f" ({fontes_list})"

cursor.execute(
    "INSERT INTO atualizacoes (tipo, descricao, qtd_inseridas) VALUES (%s, %s, %s)",
    ("carga_historica", descricao, total_inseridas)
)
conn.commit()
cursor.close()
conn.close()

print(f"\n✅ Carga histórica concluída!")
print(f"   Total encontradas: {total_encontradas}")
print(f"   Novas inseridas:   {total_inseridas}")
if por_fonte:
    print(f"   Por fonte:")
    for fonte, qtd in sorted(por_fonte.items(), key=lambda x: -x[1]):
        print(f"     {fonte}: {qtd}")
